### PR TITLE
[CDEC-658] simplify threading in block streaming

### PR DESCRIPTION
**Pull request by avieth**
## Description

If there is a connection problem causing the block streaming conversation to fail even to come up, then the retrieval worker will hang, because the block streaming `TBQueue` will never show `StreamEnd`: this is done in the `finally` clauses of the conversation callbacks _themselves_ rather than the conversation at large (they never started in this case).

The solution in this pull request is to do the queue sourcing (`retrieveBlocks`) as well as the queue sinking (`processBlocks`) from within the conversation callback itself, by way of `concurrently`. If the conversation fails to even come up, then the exception will come through, and the retrieval worker will carry on.

A bit more detail on how this arose: it required a network failure while _streaming_ blocks, rather than just fetching the next one, so it was a bit more rare than might be expected.

I'd like to make a test case for this, but it's difficult, because it requires a network failure at a particular moment. The node must be able to get the header announcements, but the network needs to go down as soon as it attempts to make a connection from streaming blocks.

The first two commits are not strictly related, but I had them around because I need them in order to build cardano-sl as a dependency. I can drop them if somebody really feels strongly that I should, but I'll need to merge them anyway.

## Linked issue

https://iohk.myjetbrains.com/youtrack/issue/CDEC-658

## Type of change

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [ ] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [ ] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.
- [ ] CHANGELOG entry has been added and is linked to the correct PR on GitHub.

## Testing checklist

- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

## How to merge

Send the message `bors r+` to merge this PR. For more information, see
[`docs/how-to/bors.md`](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/how-to/bors.md).

### CODE REVIEW

File: infra/src/Pos/Infra/Diffusion/Types.hs [hunk 1]
```diff
@@ -29,6 +31,24 @@ import           Pos.Infra.Diffusion.Subscription.Status (SubscriptionStates,
                      emptySubscriptionStates)
 import           Pos.Infra.Reporting.Health.Types (HealthStatus (..))
 
+-- | How to handle a stream of blocks.
+data StreamBlocks block m t = StreamBlocks
+  { streamBlocksMore :: NonEmpty block -> m (StreamBlocks block m t)
+    -- ^ The server gives a batch of blocks.
+  , streamBlocksDone :: m t
+    -- ^ The server has no more blocks.
+  }
```
Comment (avieth) [inline infra/src/Pos/Infra/Diffusion/Types.hs]: Introducing this type and using it in `streamBlocks` is not strictly necessary, but I had done this already for my Byron adapter branch so I included it here.

File: lib/src/Pos/Diffusion/Full/Block.hs [hunk 2]
```diff
@@ -362,57 +311,72 @@ streamBlocks logTrace smM logic streamWindow enqueue nodeId tipHeader checkpoint
...
-        :: Conc.TBQueue StreamEntry
-        -> ConversationActions MsgStream MsgStreamBlock
-        -> IO ()
-    requestBlocksConversation blockChan conv = do
+            Nothing   -> throwIO $ DialogUnexpected $ "streamBlocks: did not contact given peer"
+
+    -- The peer doesn't support streaming, we need to fall back to batching but
+    -- the current conversation is unusable since there is no way for us to learn
+    -- which blocks we shall fetch.
+    -- We will always have room to write a singel StreamEnd so there is no need to
```
Comment (coot) [inline lib/src/Pos/Diffusion/Full/Block.hs]: A typo in the line above: it should be single.

**Comment by avieth** [comment]
> I am not convinced that it will work, only because I don't see how PacketAborted get set in case of Timeout exception, maybe I missed a spot in the code, where it's done.

In the case of a `Timeout` exception, the conversation doesn't even run, and the packet can't be aborted because it's already been dequeued. Setting `PacketAborted` is an outbound queue thing, instructing it to not bother dequeueing it.

Previously we had to abort in case the processing thread (`processBlocks`) terminated before the conversation was even dequeued. Now, the processing thread won't even run until it's dequeued and active.

**Comment by coot** [comment]
> Setting PacketAborted is an outbound queue thing, instructing it to not bother dequeueing it.

This is what I cannot see, in `OutboundQueue` there is no place where `PacketAborted` is set (or anywhere else in the codebase, judging by `grep PacketAborted`.

**Comment by avieth** [comment]
>This is what I cannot see, in OutboundQueue there is no place where PacketAborted is set (or anywhere else in the codebase, judging by grep PacketAborted.

Yes it was introduced exactly for the block streaming feature. We just didn't realize there was a better way (this way) to do it.

**Pull request merged**